### PR TITLE
Support SRT styling and bold/underline in VTT

### PR DESCRIPTION
**Pull request by kloon15**
Hows this code looking like? Feedback appreciated.

### CODE REVIEW

File: srt.go [hunk 1]
```diff
@@ -111,7 +113,88 @@ func ReadFromSRT(i io.Reader) (o *Subtitles, err error) {
 			o.Items = append(o.Items, s)
 		} else {
 			// Add text
-			s.Lines = append(s.Lines, Line{Items: []LineItem{{Text: strings.TrimSpace(line)}}})
+			if l := parseTextSrt(line); len(l.Items) > 0 {
```
Comment (asticode) [inline srt.go:116]: You can replace this with `s.Lines = append(s.Lines, parseTextSrt(strings.TrimSpace(line)).Items...)`

File: srt.go [hunk 2]
```diff
@@ -111,7 +113,88 @@ func ReadFromSRT(i io.Reader) (o *Subtitles, err error) {
...
+	}
+	return
+}
+
+func parseTextSrt(i string) (o Line) {
+	// Create tokenizer
+	tr := html.NewTokenizer(strings.NewReader(i))
+
+	// Loop
+	type Styles struct {
```
Comment (asticode) [inline srt.go:130]: No need to create this struct. You can use `bold`, `italic`, `underline` and `color` variables

File: srt.go [hunk 3]
```diff
@@ -164,3 +246,67 @@ func (s Subtitles) WriteToSRT(o io.Writer) (err error) {
...
+	if b {
+		c = append(c, []byte("</b>")...)
+	}
+	if color != "" {
+		c = append(c, []byte("</font>")...)
+	}
+	return
+}
+
+func getAttribute(n *html.Token, key string) (string, bool) {
```
Comment (asticode) [inline srt.go:303]: Could you move this function to `subtitles.go`, rename it to `htmlTokenAttribute()` and sure arguments are `(t html.Token, key string) string` ?

File: srt.go [hunk 4]
```diff
@@ -111,7 +113,88 @@ func ReadFromSRT(i io.Reader) (o *Subtitles, err error) {
...
+			// Parse italic/bold/underline
+			switch token.Data {
+			case "b":
+				styles.bold = true
+			case "i":
+				styles.italic = true
+			case "u":
+				styles.underline = true
+			case "font":
+				color, _ := getAttribute(&token, "color")
```
Comment (asticode) [inline srt.go:172]: Could you replace this with

```go
if c := htmlTokenAttribute(token, "color"); c != "" {
  color = &c
}
```

File: srt.go [hunk 5]
```diff
@@ -111,7 +113,88 @@ func ReadFromSRT(i io.Reader) (o *Subtitles, err error) {
...
+				}
+			}
+		case html.TextToken:
+			if s := strings.TrimSpace(string(tr.Raw())); s != "" {
+				// Get style attribute
+				var sa *StyleAttributes
+				if styles.bold || styles.italic ||
+					styles.underline || styles.color != nil {
+					sa = &StyleAttributes{
+						TTMLColor:       styles.color,
```
Comment (asticode) [inline srt.go:184]: Could you create `StyleAttributes`'s `SRTColor`, `SRTBold`, `SRTItalic` and `SRTUnderline`, and store the values inside them instead?

File: webvtt.go [hunk 6]
```diff
@@ -290,7 +287,12 @@ func parseTextWebVTT(i string) (o Line) {
 	tr := html.NewTokenizer(strings.NewReader(i))
 
 	// Loop
-	italic := false
+	type Styles struct {
```
Comment (asticode) [inline webvtt.go:290]: Same as `srt`, no need for this struct

File: subtitles.go [hunk 7]
```diff
@@ -248,6 +250,8 @@ type StyleAttributes struct {
 	WebVTTWidth          string
 }
 
+func (sa *StyleAttributes) propagateSRTAttributes() {}
```
Comment (asticode) [inline subtitles.go:253]: You need copy new `SRT*` attributes into proper fields such as corresponding `WebVTT*` fields. 

Same in `propagateWebVTTAttributes`, you need to do the reverse

**Review by asticode — CHANGES_REQUESTED** [review]
Could you also fix webvtt tests and add srt/webvtt tests for bold/italic/underline/color?

You can either directly use `testdata/example-in.srt` in `TestSRT` for instance, or write a new test specifically for html tags.

File: srt.go (hunk 1 above)
Comment (kloon15) [reply srt.go:116]: Thats no bueno, i think u want this?
`s.Lines = append(s.Lines, parseTextSrt(line))`
I just straight up copied this part from the webvtt equivalent, should i change it there too then?

Comment (asticode) [reply srt.go:116]: You're right, just replace `line` with `strings.TrimSpace(line)` only here, no need to update the webvtt counterpart

Comment (kloon15) [reply srt.go:116]: Thats done inside the function like in webvtt.

**Comment by asticode** [comment]
Could you remove the SSA-related commits from this PR?

I don't want to mix different things in the same PR.

Thanks

**Comment by kloon15** [comment]
> Could you remove the SSA-related commits from this PR?
> 
> I don't want to mix different things in the same PR.
> 
> Thanks

I done nothing to SSA, its for supporting SRT position tags which use ASS tags so figured its easier to call em that.

**Comment by asticode** [comment]
> I done nothing to SSA, its for supporting SRT position tags which use ASS tags so figured its easier to call em that. TBH im not sure what tags most players support, but position is pretty popular.

OK, then could you remove the 2 commits handling ASS tags from this PR? I'd rather have this change in a separate PR.

Thanks!

**Comment by kloon15** [comment]
Sure will do that.

**Comment by asticode** [comment]
It's been over a year, let me know whether this PR is still active 👍

**Comment by kloon15** [comment]
No clue, I got busy with life, so idk what happened here since, whether its still relevant.

**Comment by asticode** [comment]
No worries, thanks again ❤️

**Pull request closed without merging**